### ISSUE TITLE
trainable=False on shared layers broken in 2.2.4-tf?

### ISSUE THREAD

**Issue by sandeepjana** [keras-team/keras#13436]
My model contains shared layers that are wrapped by (sub-)models. Setting `trainable` flag on one sub-model is causing the layers themselves to freeze, which is unexpected. This issue seems to appear only in keras version 2.2.4-tf (tf 1.14). 

Sample:
```
x = layers.Input((1,))
linear = layers.Dense(1)
frozen = Model(x, linear(x))
frozen.trainable = False
print(frozen.trainable, linear.trainable) 
# Output: False False. 
# I was expecting False True (and indeed that's the output in version 2.3.1)

```
As per my understanding, setting trainable flag on a model does not change the trainable property of the underlying layers as alluded by @fchollet here: https://github.com/keras-team/keras/issues/4674#issuecomment-293362106

**System information**  
- Have I written custom code (as opposed to using example directory):  yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Windows 10
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  1.14
- Keras version:  2.2.4-tf
- Python version:  3.7.4
- CUDA/cuDNN version: NA  
- GPU model and memory: NA 

---------

After a little more testing, it seems none of the keras versions I tested gave the results I expected when it comes to model training behavior.  

The setup below is like this: the model is a multi output classification network with one branch set to trainable and another branch frozen -- yet both share weights. I train the model with (1) same set of labels for both outputs (2) one output has labels reversed. 

I was expecting the model would fit the data with original labels (for the trainable branch) in both scenarios as the frozen branch would not play a role in training process. That is not what happened.

```
import numpy as np
import tensorflow.keras as keras
from tensorflow.keras import layers, optimizers
from tensorflow.keras.models import Model
# import keras
# from keras import layers, optimizers
# from keras.models import Model

print('version', keras.__version__)

def model(name):
    x = layers.Input((1,))
    y = layers.Dense(1, activation='sigmoid')(x)
    return Model(inputs=x, outputs=y, name=name)

trainable = model('trainable')
frozen = Model(inputs=trainable.input, outputs=trainable.output, name='frozen')
frozen.trainable = False

x = layers.Input((1,))
combined = Model(inputs=x, outputs=[trainable(x), frozen(x)])

optim = optimizers.SGD(lr=5e-1)
combined.compile(loss=['binary_crossentropy', 
    'binary_crossentropy'], optimizer=optim)

print(trainable.trainable, frozen.trainable)
print([a.trainable for a in trainable.layers[1:]])
print([a.trainable for a in frozen.layers[1:]])
print([(a.name, a.trainable) for a in combined.layers[1:]])

n = 64
X = np.ones((n, 1))
X[:n//2, 0] = -1
t = 1
Y = np.ones((n, 1)) * t
Y[:n//2, 0] = (1 - t)
Z = np.ones((n, 1)) * t
Z[n//2:, 0] = (1 - t)

indices = np.random.permutation(n)[:1]

for i, aux in enumerate([Y, Z]):
    combined.fit(X, [Y, aux], epochs=20, verbose=0, validation_split=0.1)
    predictions = combined.predict(X)
    assert(np.all(predictions[0] == predictions[1]))

    print(i, np.hstack([predictions[0][indices], predictions[1][indices], Y[indices]]))

```

Outputs with different versions of keras:

```
version 2.2.4-tf (tf 1.14)
True False
[False]
[False]
[('trainable', True), ('frozen', False)]
0 [[0.67681402 0.67681402 1.        ]]
1 [[0.67681402 0.67681402 1.        ]]

version 2.1.6-tf (tf 1.12)
True False
[True]
[True]
[('trainable', True), ('frozen', False)]
0 [[0.02213766 0.02213766 0.        ]]
1 [[0.4999963 0.4999963 0.       ]]

version 2.3.1
True False
[True]
[True]
[('trainable', True), ('frozen', False)]
0 [[0.97395408 0.97395408 1.        ]]
1 [[0.50008553 0.50008553 1.        ]]

version 2.2.4 (tf 1.12)
True False
[True]
[True]
[('trainable', True), ('frozen', False)]
0 [[0.0230438 0.0230438 0.       ]]
1 [[0.4999963 0.4999963 0.       ]]

```

**Comment by sandeepjana** [comment]
I think I have an explanation for why frozen branch is playing role in the training process (*second* part of above post): the same weight tensor is used in computation of total loss that includes losses from both branches. During the differentiation step, there is no way to tell apart the loss contribution from the frozen and trainable branches if the weight is shared. 

Some details:
(1) Total loss tensor is computed by summing up the contributions from both trainable and frozen branches: `L = f(w1, ...) + g(w2, ...)` where `w1`, `w2` are weights, and f and g are transformations of individual branches. 
(2) If both branches are trainable, then the trainable weights list is `[w1, w2]`. If `w2` is on a frozen branch, then it is not included in the trainable weights. In this case, trainable weights is just `[w1]`. This implies `dL/dw1 = f'(w1)` and `dL/dw2 = 0`. The weight `w1` is updated based on `f'(w1)` component. and there is no weight update for `w2`. 
(3) However, if `w1` and `w2` are the shared weights just being aliases for weight `w`, then the trainable weights list is `[w]` when one branch is frozen. (It would have been `[w, w]` if both branches are trainable). In this case, `dL/dw = f'(w) + g'(w)`. I expected the weight update from `f'(w)` component only in the above post, but that is not the case.

The graph should treat weight on frozen branch as a `constant` rather than a `variable`.

**Comment by edmondja** [comment]
I think I have the same problem but I think I found a solution here.
https://github.com/keras-team/keras/issues/8259
